cpp/src_prims/matrix/batched_matrix.hpp

```diff
@@ -0,0 +1,445 @@
+    BatchedMatrix toMat(m, n, m_num_batches, m_cublasHandle, m_allocator,
+                        m_stream);
+    cudaMemcpyAsync(toMat[0], this->operator[](0),
+                    m_num_batches * r * sizeof(double),
+                    cudaMemcpyDeviceToDevice, m_stream);
+
+    return toMat;
+  }
+
+  //! Visualize the first matrix.
+  void print(std::string name) const {
+    size_t len = m_shape.first * m_shape.second * m_num_batches;
+    std::vector<double> A(len);
+    updateHost(A.data(), m_A_dense.get(), len, m_stream);
+    std::cout << name << "=\n";
+    for (int i = 0; i < m_shape.first; i++) {
+      for (int j = 0; j < m_shape.second; j++) {
+        // column major
+        std::cout << std::setprecision(10) << A[j * m_shape.first + i] << ",";
+      }
+      std::cout << "\n";
+    }
+  }
+
+  /* Initialize a batched identity matrix.
+   * @param m Number of rows/columns of matrix
+   * @param num_batches Number of matrices in batch
+   * @param pool Memory pool
+   */
+  static BatchedMatrix Identity(int m, int num_batches, cublasHandle_t handle,
+                                std::shared_ptr<ML::deviceAllocator> allocator,
+                                cudaStream_t stream) {
+    BatchedMatrix I(m, m, num_batches, handle, allocator, stream, true);
+
+    identity_matrix_kernel<<<num_batches, std::min(1024, m), 0, stream>>>(
+      I.data(), m);
+    CUDA_CHECK(cudaGetLastError());
+    return I;
+  }
+
+ private:
+  //! Shape (rows, cols) of matrices. We assume all matrices in batch have same shape.
+  std::pair<int, int> m_shape;
+
+  //! Array(pointer) to each matrix.
+  std::shared_ptr<double*> m_A_batches;
+
+  //! Data pointer to first element of dense matrix data.
+  std::shared_ptr<double> m_A_dense;
+
+  //! Number of matrices in batch
+  size_t m_num_batches;
+
+  std::shared_ptr<deviceAllocator> m_allocator;
+  cublasHandle_t m_cublasHandle;
+  cudaStream_t m_stream;
+};
+
+/* Computes batched kronecker prodcut between AkB <- A (x) B
+ * @param A Array of pointers to matrices `A`
+ * @param m number of rows (A)
+ * @param n number of columns (A)
+ * @param B Array of pointers to matrices `B`
+ * @param p number of rows (B)
+ * @param q number of columns (B)
+ * @param AkB Result kronecker product pointer array.
+ * @param k_m Result number of rows
+ * @param k_n Result number of columns
+ */
+__global__ void kronecker_product_kernel(double** A, int m, int n, double** B,
+                                         int p, int q, double** AkB, int k_m,
+                                         int k_n) {
+  int bid = blockIdx.x;
+
+  for (int ia = 0; ia < n; ia++) {
+    for (int ja = 0; ja < m; ja++) {
+      double A_ia_ja = A[bid][ia + ja * m];
+
+      for (int iblock = 0; iblock < p; iblock += blockDim.x) {
+        int ib = threadIdx.x + iblock * blockDim.x;
+        if (ib < p) {
+          for (int jblock = 0; jblock < q; jblock += blockDim.y) {
+            int jb = threadIdx.y + jblock * blockDim.y;
```

Comment:
I think that these lines are wrong: the increment in the loops should be 1, since the multiplication by `blockDim.x` or `blockDim.y` is done in the calculation of `ib` or `jb`.

E.g imagine that you have `p=8` and `blockDim.x=4`. Here is the comparison of the current values that we get for `ib` against the ones that we should get:

|thread id|actual|correct|
|---|---|---|
|0|0, 16|0, 4|
|1|1, 17|1, 5|
|2|2, 18|2, 6|
|3|3, 19|3, 7|

The unit tests are too small to catch the problem, as p and q are smaller than 32 so there will always be the need of only one iteration of these two loops per thread, and the first value is right. This means that the **unit tests will need to be improved to be more generic.**

I will fix this issue while working on #1227

Reply:
ah.. good catch. Thanks @nyrio!

Reply:
Solved, discussion can be closed